### PR TITLE
Feat: admin withdraw + claim deadline

### PR DESCRIPTION
**Pull request by sssrot**

### CODE REVIEW

File: programs/merkle-distributor/src/lib.rs [hunk 1]
```diff
@@ -171,6 +180,41 @@ pub mod merkle_distributor {
...
+        }
+
+        let config = &mut ctx.accounts.config;
+        config.distributor = ctx.accounts.distributor.key(); //init for empty config
+        config.claim_deadline = claim_deadline;
+
+        Ok(())
+    }
+
+    #[allow(deprecated)]
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs]: 为什么这里是 allow deprecated？

File: programs/merkle-distributor/src/lib.rs [hunk 2]
```diff
@@ -294,6 +375,23 @@ pub struct MerkleDistributor {
     pub num_nodes_claimed: u64,
 }
 
+#[account]
+#[derive(Default)]
+pub struct Config {
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs:387]: did we initiate merkle distributor with extra space s.t. we can add these directly to the distributor?

File: programs/merkle-distributor/src/lib.rs [hunk 3]
```diff
@@ -83,6 +83,15 @@ pub mod merkle_distributor {
         amount: u64,
         proof: Vec<[u8; 32]>,
     ) -> ProgramResult {
+        if **ctx.accounts.config.lamports.borrow() > 0u64 {
+            let config: Account<Config> = Account::try_from(&ctx.accounts.config)?;
+            require!(
+                config.distributor == ctx.accounts.distributor.key(),
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs]: 为什么这里还要检查一次？好像 `Context<Claim>` 已经检查了。

这里只要检查 config 有没有被 initialize 过就行了？

File: programs/merkle-distributor/src/lib.rs [hunk 4]
```diff
@@ -83,6 +83,15 @@ pub mod merkle_distributor {
         amount: u64,
         proof: Vec<[u8; 32]>,
     ) -> ProgramResult {
+        if **ctx.accounts.config.lamports.borrow() > 0u64 {
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs]: 我觉得抽出一个函数：

```
// Load an account that may be empty. If it's empty return None.
fn load_optional_account<T>(account) -> Option<T> {
  if ctx.accounts.config.lamports == 0 {
     return None;
  }
  let acc: Account<t> = Account::try_from(accounts)?;  
  Some(acc)
}
```

```
if Some(config) = load_optional_account(&ctx.accounts.config) {
}
```

**Review by defactojob — CHANGES_REQUESTED** [review]

File: programs/merkle-distributor/src/lib.rs (hunk 1 above)
Comment (sssrot) [reply programs/merkle-distributor/src/lib.rs]: Removed.

File: programs/merkle-distributor/src/lib.rs (hunk 2 above)
Comment (sssrot) [reply programs/merkle-distributor/src/lib.rs:387]: Yeah, init_if_need space 256 when UpdateConfig

File: programs/merkle-distributor/src/lib.rs (hunk 3 above)
Comment (sssrot) [reply programs/merkle-distributor/src/lib.rs]: 对，其实不检查也没问题，因为用seed限制了config 和distributor的 1:1 bind，算是double check吧

Comment (defactojob) [reply programs/merkle-distributor/src/lib.rs]: 没有必要的话咋们还是移除这里的额外检查吧。

context 检查过的就不检查了，要不然有点黑人问号。。。？

File: programs/merkle-distributor/src/lib.rs [hunk 5]
```diff
@@ -294,6 +383,23 @@ pub struct MerkleDistributor {
     pub num_nodes_claimed: u64,
 }
 
+#[account]
+#[derive(Default)]
+pub struct Config {
+    pub distributor: Pubkey,
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs:388]: do we need config.distributor?

Comment (defactojob) [reply programs/merkle-distributor/src/lib.rs:388]: i guess having it is slightly more convenient if we wanna dump and correlate the data

File: programs/merkle-distributor/src/lib.rs [hunk 6]
```diff
@@ -171,6 +189,40 @@ pub mod merkle_distributor {
 
         Ok(())
     }
+    pub fn update_config(ctx: Context<UpdateConfig>, claim_deadline: Option<i64>) -> ProgramResult {
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs:188]:
```
    pub fn update_config(ctx: Context<UpdateConfig>, claim_deadline: Option<i64>) -> ProgramResult {
        let config = &mut ctx.accounts.config;
        config.distributor = ctx.accounts.distributor.key(); //init for empty config

       if Some(claim_deadline) = claim_deadline {
            require!(claim_deadline.unwrap() >= 0, InvalidParams);
            config.claim_deadline = claim_deadline;
        }


        

        Ok(())
    }
```

File: programs/merkle-distributor/src/lib.rs [hunk 7]
```diff
@@ -22,6 +22,16 @@ pub mod merkle_proof;
 
 declare_id!("PMRKTWvK9f1cPkQuXvvyDPmyCSoq8FdedCimXrXJp8M");
 
+// Load an account that may be empty. If it's empty return None.
+fn load_optional_account<'a, T: AccountSerialize + AccountDeserialize + Owner + Clone>(
+    info: &AccountInfo<'a>,
+) -> Result<Option<Account<'a, T>>> {
+    if **info.lamports.borrow() == 0u64 {
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs]: could there be a DOS attack, where i send 0.001 SOL to the expected address, and this function would always fail?

maybe it's better to check if the account data has 0 byte?

**Review by defactojob — CHANGES_REQUESTED** [review]

File: programs/merkle-distributor/src/lib.rs [hunk 8]
```diff
@@ -22,6 +22,16 @@ pub mod merkle_proof;
 
 declare_id!("PMRKTWvK9f1cPkQuXvvyDPmyCSoq8FdedCimXrXJp8M");
 
+// Load an account that may be empty. If it's empty return None.
+fn load_optional_account<'a, T: AccountSerialize + AccountDeserialize + Owner + Clone>(
+    info: &AccountInfo<'a>,
+) -> Result<Option<Account<'a, T>>> {
+    if **info.lamports.borrow() == 0u64 || info.data_len() == 0 {
```
Comment (defactojob) [inline programs/merkle-distributor/src/lib.rs:29]: how about `**info.lamports.borrow() == 0u64 && info.data_len() == 0`

**Review by defactojob — CHANGES_REQUESTED** [review]

**Pull request merged**